### ISSUE TITLE
gRPC requests and per try timeout

### ISSUE THREAD

**Issue by snowp** [envoyproxy/envoy#4628]
From reading the source, it seems like when `max_grpc_timeout` is set we use `grpc-timeout` as the global timeout, but if we additionally specify a `per_retry_timeout` then the upstream will see the global timeout as `grpc-timeout` even though Envoy will reset the stream after the per try timeout has been reached.

If I'm understanding this correctly, then it seems to me like `grpc-timeout` should be set to the retry timeout (not the global timeout) so that servers are aware of how much time remains before the request will be terminated by the client (Envoy). Alternatively I could have all my servers look for both `grpc-timeout` and `x-envoy-expected-rq-timeout-ms`, but it seems unnecessary to have to use the Envoy header when there already exists a gRPC specific timeout header. 

Thoughts?

**Comment by snowp** [comment]
Not quite, what I'm saying is that if you send a gRPC request like this: 

```
ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
defer cancel()

ctx = metadata.AppendToOutgoingContext(ctx, "x-envoy-upstream-rq-per-try-timeout-ms", "100")
r, err := c.SendData(ctx, &service.HiRequest{})
```

then with the following HCM config:

```
      - name: envoy.http_connection_manager
        config:
          codec_type: auto
          stat_prefix: ingress_http
          use_remote_address: true
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains:
              - "*"
              routes:
              - match:
                  prefix: "/"
                route:
                  cluster: app
                  max_grpc_timeout: 0s
          http_filters:
          - name: envoy.router
```

you end up setting `grpc-timeout` much higher than the actual timeout imposed by Envoy:

```
'grpc-timeout', '4999971u'
'x-forwarded-for', '172.17.0.1'
'x-forwarded-proto', 'http'
'x-envoy-internal', 'true'
'x-request-id', 'edff67ed-1d79-470c-a794-9436f104f0b7'
'x-envoy-expected-rq-timeout-ms', '100'
```

This means that an upstream gRPC server would look at it's gRPC timeout and think that it has 5 seconds to complete the request even though Envoy will fail the request with a timeout in 100ms.

Basically what I'm asking is: is it reasonable to rewrite `grpc-timeout` to match `x-envoy-expected-rq-timeout-ms`?

**Comment by mattklein123** [comment]
Ah, yes. I see. Yes that clearly seems like a bug. Switching the label over. Do you want to fix?

**Comment by snowp** [comment]
Yeah I’ll take it
